File: morning_exercises/day16.cpp
#include <climits>
#include <iostream>
#include <memory>
#include <vector>

class BinaryNode {
    public:
        int val = 0;
        std::shared_ptr<BinaryNode> left = nullptr;
        std::shared_ptr<BinaryNode> right = nullptr;
    BinaryNode(int val): val(val) {}

    static void goodNodesMain(std::shared_ptr<BinaryNode>& root, std::vector<int>& result, int min) {
        if (root == nullptr) return;

        // if (root->val<min) return;

        if (root->val>=min) result.push_back(root->val);
        goodNodesMain(root->left, result, std::max(root->val, min));
        goodNodesMain(root->right, result, std::max(root->val, min));
    }

    static std::vector<int> goodNodes(std::shared_ptr<BinaryNode> root) {
        std::vector<int> result = {};
        
        BinaryNode::goodNodesMain(root, result, INT_MIN);

        return result;
    }
};

int main() {
    auto root = std::make_shared<BinaryNode>(5);
    root->left = std::make_shared<BinaryNode>(3);
    root->left->left = std::make_shared<BinaryNode>(1);
    root->left->right = std::make_shared<BinaryNode>(4);
    root->right = std::make_shared<BinaryNode>(2);
    root->right->right = std::make_shared<BinaryNode>(7);
    root->right->left = std::make_shared<BinaryNode>(1);

    std::vector<int> result = BinaryNode::goodNodes(root);
    for (int c : result) {
        std::cout << c << "\n";
    }
}

File: morning_exercises/day16.py
from __future__ import annotations

class BinaryNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

    @staticmethod
    def isGoodNode(root: BinaryNode | None, result: list, _max: float)->None:
        if root is None: return

        if (root.val>_max): result.append(root.val)
        BinaryNode.isGoodNode(root.left, result, max(_max, root.val))
        BinaryNode.isGoodNode(root.right, result, max(_max, root.val))

    @staticmethod
    def goodNodes(root: BinaryNode | None)->list:
        result = []
        BinaryNode.isGoodNode(root, result, float('-inf'))
        return result

root = BinaryNode(6)
root.left = BinaryNode(8)
root.right = BinaryNode(3)
root.left.left = BinaryNode(2)
root.left.right = BinaryNode(5)
root.right.left = BinaryNode(7)
root.right.right = BinaryNode(9)

print(BinaryNode.goodNodes(root))